uaclient/entitlements/fips.py

```diff
@@ -62,6 +62,41 @@ def conditional_packages(self):
 
         return conditional_packages
 
+    def install_packages(
+        self,
+        package_list: "List[str]" = None,
+        cleanup_on_failure: bool = True,
+        verbose: bool = True,
+    ) -> None:
+        """Install contract recommended packages for the entitlement.
+
+        :param package_list: Optional package list to use instead of
+            self.packages.
+        :param package_list: Optional package list to use instead of
+            self.packages.
+        """
+        print("Installing {title} packages".format(title=self.title))
+
+        # We need to guarantee that the metapackage is installed.
+        # While the other packages should still be installed, if they
+        # fail, we should not block the enable operation.
+        fips_metapackage = self.packages[0]
```

Comment:
This seems very brittle, if the contract ever changes the number of mandatory additionalPackages directives to something greater than one, we now will treat those packages as optional too.  I think we need to treat conditional packages uniquely from the additionalPackage directives since they are a different "type" of directive that we treat like  `addPackagesAllowInstallFailures` if the contract server we able to surface this type of directive.


What do you think about trimming down  FIPS.packages method to only fixup the proper cloud metapkg name and just treating the conditional_packages here in your sub-classed install_packages method?

I was thinking about something like this
```diff
diff --git a/uaclient/entitlements/fips.py b/uaclient/entitlements/fips.py
index 60170a17..e9692b7b 100644
--- a/uaclient/entitlements/fips.py
+++ b/uaclient/entitlements/fips.py
@@ -78,14 +78,27 @@ class FIPSCommonEntitlement(repo.RepoEntitlement):
         print("Installing {title} packages".format(title=self.title))
 
         # We need to guarantee that the metapackage is installed.
-        # While the other packages should still be installed, if they
-        # fail, we should not block the enable operation.
+        # The mandatory meta package is present in addtionalPackages directive
+        # and can be installed by the default install_packages method.
         fips_metapackage = self.packages[0]
-        super().install_packages(
-            package_list=[fips_metapackage], verbose=False
+        super().install_packages(verbose=False)
+
+        # Any conditional packages packages should still be installed, but if
+        # they fail to install we should not block the enable operation.
+        installed_packages = apt.get_installed_packages()
+
+        desired_packages = []
+        pkg_groups = groupby(
+            sorted(self.conditional_packages),
+            key=lambda pkg_name: pkg_name.replace("-hmac", ""),
         )
 
-        for pkg in self.packages[1:]:
+        for pkg_name, pkg_list in pkg_groups:
+            if pkg_name in installed_packages:
+                desired_packages += pkg_list
+
+
+        for pkg in desired_packages:
             try:
                 super().install_packages(
                     package_list=[pkg], cleanup_on_failure=False, verbose=False
@@ -220,17 +233,6 @@ class FIPSCommonEntitlement(repo.RepoEntitlement):
     def packages(self) -> "List[str]":
         packages = super().packages
         packages = self._replace_metapackage_on_cloud_instance(packages)
-        installed_packages = apt.get_installed_packages()
-
-        pkg_groups = groupby(
-            sorted(self.conditional_packages),
-            key=lambda pkg_name: pkg_name.replace("-hmac", ""),
-        )
-
-        for pkg_name, pkg_list in pkg_groups:
-            if pkg_name in installed_packages:
-                packages += pkg_list
-
         return packages
 
     def application_status(self) -> "Tuple[status.ApplicationStatus, str]":
```

Reply:
If we're handling errors when installs fail, do we even need to do the `desired_packages` logic (to only try to upgrade if already installed) anymore?

Or with this new approach, can we just iterate over `conditional_packages` directly, try to install each one individually, and catch any errors?

Reply:
I think we should because otherwise we could end up unnecessary packages in the system. For example, if the user doesn't have `strongswan` in the system, we don't need to install the FIPS version of it for the system to be FIPS compliant.

Reply:
Good point! Thats good with me